netperf_tool/src/client_node.cpp

```diff
@@ -0,0 +1,160 @@
+// Copyright 2022 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "client_node.hpp"
+
+#include <chrono>
+#include <functional>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+
+#include "netperf_tool_interfaces/msg/bytes.hpp"
+#include "netperf_tool_interfaces/srv/get_results.hpp"
+
+#include "rclcpp/executors.hpp"
+#include "rclcpp/node.hpp"
+#include "rclcpp/serialization.hpp"
+#include "rclcpp/serialized_message.hpp"
+
+#include "utils.hpp"
+
+namespace netperf_tool
+{
+
+using namespace std::chrono_literals;
+
+ClientNode::ClientNode(
+  size_t array_size,
+  std::chrono::nanoseconds target_pub_period,
+  const rclcpp::QoS & pub_qos,
+  const rclcpp::NodeOptions & options)
+: Node("client", "netperf_tool", options),
+  array_size_{array_size},
+  target_pub_period_{target_pub_period}
+{
+  pub_ = this->create_publisher<netperf_tool_interfaces::msg::Bytes>("test_topic", pub_qos);
+  // always keep the vector preallocated, to avoid delays when the timer is triggered
+  bytes_.resize(array_size);
+  client_ = this->create_client<netperf_tool_interfaces::srv::GetResults>("get_results");
+}
+
+bool
+ClientNode::wait_for_server(std::chrono::nanoseconds timeout)
+{
+  auto start = std::chrono::system_clock::now();
+  while (std::chrono::system_clock::now() < start + timeout) {
+    if (client_->service_is_ready()) {
+      return true;
+    }
+    RCLCPP_INFO(this->get_logger(), "perf server not available yet, waiting...");
+    std::this_thread::sleep_for(500ms);
+  }
+  RCLCPP_INFO_STREAM(
+    this->get_logger(),
+    "perf server not available after waiting ["
+      << std::chrono::duration_cast<std::chrono::seconds>(timeout).count() << "s]" << std::endl);
+  return false;
+}
+
+void
+ClientNode::start_publishing()
+{
+  timer_ = this->create_wall_timer(
+    target_pub_period_,
+    std::bind(&ClientNode::pub_next_msg, this));
+}
+
+void
+ClientNode::stop_publishing()
+{
+  if (timer_) {
+    timer_->cancel();
+  }
+}
+
+void
+ClientNode::sync_with_server(rclcpp::Executor & exec)
+{
+  if (!pub_->wait_for_all_acked(5s)) {
+    RCLCPP_WARN(this->get_logger(), "Some messages were not acked by the perf server ...");
```

Comment:
I think you should tidy up the terminology in these messages, "perf server" might be confusing to someone, instead consider using "netperf server", as that is what the users actually run (`ros2 netperf server`).

Reply:
nitpick:

```suggestion
    RCLCPP_WARN(this->get_logger(), "Some messages were not acknowledged by the perf server...");
```

Reply:
> I think you should tidy up the terminology in these messages, "perf server" might be confusing to someone, instead consider using "netperf server", as that is what the users actually run (ros2 netperf server).

Yes, I started calling the packages `perf_tool/ros2perf` and later renamed them to `netperf_tool/ros2netperf`.
It seems that I forgot to replace the old name for the new one in some places.

```suggestion
    RCLCPP_WARN(this->get_logger(), "Some messages were not acknowledged by the netperf server ...");
```